server/src/main/java/org/apache/iotdb/db/mpp/memory/MemoryPool.java

```diff
@@ -133,6 +133,31 @@ public boolean tryReserve(String queryId, long bytes) {
     return true;
   }
 
+  /**
+   * Cancel the specified memory reservation. If the reservation has finished, do nothing.
+   *
+   * @param future The future returned from {@link #reserve(String, long)}
+   * @return If the future has not complete, return the number of bytes being reserved. Otherwise,
+   *     return 0.
+   */
+  public synchronized long tryCancel(ListenableFuture<Void> future) {
+    Validate.notNull(future);
+    if (future.isDone()) {
+      return 0L;
+    }
+    Iterator<MemoryReservationFuture<Void>> iterator = memoryReservationFutures.iterator();
+    while (iterator.hasNext()) {
+      MemoryReservationFuture<Void> f = iterator.next();
+      if (future.equals(f)) {
+        f.cancel(true);
+        iterator.remove();
+        return f.getBytes();
+      }
```

Comment:
Here, time complexity of removing is O(N), but we may not need to do real removing here. We can just cancel it. Removing job can be left to free method

Reply:
Make sense. Fixed.